storage/inkless/src/test/java/io/aiven/inkless/consume/FetchPlannerTest.java

```diff
@@ -210,19 +225,391 @@ private FetchPlanner fetchPlannerJob(Map<TopicIdPartition, FindBatchResponse> ba
+                new BatchInfo(2L, OBJECT_KEY_A.value(),
+                    BatchMetadata.of(partition1, 30, 10, 0, 0, 15, 500, TimestampType.CREATE_TIME))
+            ), 0, 1)
+        );
+
+        final FetchPlanner planner = fetchPlannerJob(coordinates);
+        final List<ObjectFetchRequest> result = planner.planJobs(coordinates);
+
+        // Should merge into single request for OBJECT_KEY_A
+        assertThat(result).hasSize(1);
+
+        final ObjectFetchRequest request = result.get(0);
+        assertThat(request.objectKey()).isEqualTo(OBJECT_KEY_A);
+        // Timestamp should be max(100, 500) = 500
+        assertThat(request.timestamp()).isEqualTo(500L);
+        // ByteSize should be sum(10, 10) = 20
+        assertThat(request.byteSize()).isEqualTo(20L);
+    }
+
+    @Test
+    public void testConcurrentRequestsToSameKeyFetchOnlyOnce() throws Exception {
+        // Test that when multiple requests for the SAME cache key arrive concurrently,
+        // the underlying fetch operation is performed only ONCE, and all requesters
+        // receive the same result.
+        //
+        // This validates the cache's deduplication behavior which prevents redundant
+        // fetches to object storage when multiple threads/requests need the same data.
+
+        try (CaffeineCache caffeineCache = new CaffeineCache(100, 3600, 180)) {
+            final byte[] expectedData = "shared-data".getBytes();
+
+            // Mock fetcher to return data
+            when(fetcher.fetch(eq(OBJECT_KEY_A), any(ByteRange.class)))
+                .thenReturn(null);
+            when(fetcher.readToByteBuffer(any()))
+                .thenReturn(ByteBuffer.wrap(expectedData));
+
+            // Create coordinates with TWO batches that map to the SAME cache key
+            // (same object, same byte range after alignment)
+            final Map<TopicIdPartition, FindBatchResponse> coordinates = Map.of(
+                partition0, FindBatchResponse.success(List.of(
+                    new BatchInfo(1L, OBJECT_KEY_A.value(),
+                        BatchMetadata.of(partition0, 0, 10, 0, 0, 10, 20, TimestampType.CREATE_TIME))
+                ), 0, 1),
+                partition1, FindBatchResponse.success(List.of(
+                    new BatchInfo(2L, OBJECT_KEY_A.value(),
+                        BatchMetadata.of(partition1, 0, 10, 0, 0, 10, 20, TimestampType.CREATE_TIME))
+                ), 0, 1)
+            );
+
+            final FetchPlanner planner = new FetchPlanner(
+                time, OBJECT_KEY_CREATOR, keyAlignmentStrategy,
+                caffeineCache, fetcher, dataExecutor, coordinates, metrics
+            );
+
+            // Execute: Both batches will create fetch requests for the same cache key
+            final List<CompletableFuture<FileExtent>> futures = planner.get();
+
+            // Should have only 1 future because the cache deduplicates same-key requests
+            assertThat(futures).hasSize(1);
```

Comment:
This will happen not because of the cache key but because of `groupingBy` with `BatchInfo.objectKey`. There will be only single FetchRequest created always because of that. Might be so that this is still fine because the idea is to test deduplication but the comment seems to be misleading.